File: fluid_simulation/fluid.py
import math
from functools import lru_cache
from typing import ClassVar


class Fluid:
    N: ClassVar[int] = 64  # size (how many squares)
    iterations: ClassVar[int] = 10

    def __init__(self, dt: float, diffussion: float, viscosity: float) -> None:
        self.dt = dt
        self.diff = diffussion
        self.visc = viscosity

        self.s: list[float] = [0] * (self.N * self.N)
        self.density: list[float] = [0] * (self.N * self.N)

        self.Vx: list[float] = [0] * (self.N * self.N)
        self.Vy: list[float] = [0] * (self.N * self.N)

        self.Vx0: list[float] = [0] * (self.N * self.N)
        self.Vy0: list[float] = [0] * (self.N * self.N)

    def add_density(self, x: int, y: int, amount: int) -> None:
        """Add density to the element that will be added to the water (e.g. soy sauce)"""
        index = IX(x, y)
        self.density[index] += amount

    def add_velocity(self, x: int, y: int, amount_x: float, amount_y: float) -> None:
        index = IX(x, y)
        self.Vx[index] += amount_x
        self.Vy[index] += amount_y

    def step(self) -> None:
        """Go through time with the fluid"""

        diffuse(1, self.Vx0, self.Vx, self.visc, self.dt, self.iterations, self.N)
        diffuse(2, self.Vy0, self.Vy, self.visc, self.dt, self.iterations, self.N)

        project(self.Vx0, self.Vy0, self.Vx, self.Vy, self.iterations, self.N)

        advect(1, self.Vx, self.Vx0, self.Vx0, self.Vy0, self.dt, self.N)
        advect(2, self.Vy, self.Vy0, self.Vx0, self.Vy0, self.dt, self.N)

        project(self.Vx, self.Vy, self.Vx0, self.Vy0, self.iterations, self.N)

        diffuse(0, self.s, self.density, self.diff, self.dt, self.iterations, self.N)
        advect(0, self.density, self.s, self.Vx, self.Vy, self.dt, self.N)

    def fade_density(self, amount):
        for i in range(0, len(self.density)):
            if self.density[i] > 0:
                self.density[i] -= amount


def diffuse(
    b: int, x: list[float], x0: list[float], diff: float, dt: float, iter: int, N: int
) -> None:
    a: float = dt * diff * (N - 2) * (N - 2)
    lin_solve(b, x, x0, a, 1 + 6 * a, iter, N)


@lru_cache(maxsize=Fluid.N * Fluid.N)
def constrain(value, minimum, maximum):
    if value < minimum:
        return minimum
    elif value > maximum:
        return maximum
    else:
        return value


@lru_cache(maxsize=Fluid.N * Fluid.N)
def IX(
    x: int, y: int, N: int = Fluid.N
) -> int:  # TODO: Fix value for N (should be dynamic)
    """Return 2D location as a 1D index"""
    x = constrain(x, 0, N - 1)
    y = constrain(y, 0, N - 1)
    return int(x + y * N)


def lin_solve(
    b: int, x: list[float], x0: list[float], a: float, c: float, iter: int, N: int
) -> None:
    cRecip: float = 1.0 / c
    for _ in range(0, iter):
        for j in range(1, N - 1):
            for i in range(1, N - 1):
                x[IX(i, j)] = (
                    x0[IX(i, j)]
                    + a
                    * (
                        x[IX(i + 1, j)]
                        + x[IX(i - 1, j)]
                        + x[IX(i, j + 1)]
                        + x[IX(i, j - 1)]
                    )
                    * cRecip
                )
        set_bnd(b, x, N)


def project(
    velocX: list[float],
    velocY: list[float],
    p: list[float],
    div: list[float],
    iter: int,
    N: int,
) -> None:
    for j in range(1, N - 1):
        for i in range(1, N - 1):
            div[IX(i, j)] = (
                -0.5
                * (
                    velocX[IX(i + 1, j)]
                    - velocX[IX(i - 1, j)]
                    + velocY[IX(i, j + 1)]
                    - velocY[IX(i, j - 1)]
                )
                / N
            )
            p[IX(i, j)] = 0

    set_bnd(0, div, N)
    set_bnd(0, p, N)
    lin_solve(0, p, div, 1, 6, iter, N)

    for j in range(1, N - 1):
        for i in range(1, N - 1):
            velocX[IX(i, j)] -= 0.5 * (p[IX(i + 1, j)] - p[IX(i - 1, j)]) * N
            velocY[IX(i, j)] -= 0.5 * (p[IX(i, j + 1)] - p[IX(i, j - 1)]) * N
    set_bnd(1, velocX, N)
    set_bnd(2, velocY, N)


def advect(
    b: int,
    d: list[float],
    d0: list[float],
    velocX: list[float],
    velocY: list[float],
    dt: float,
    N: int,
) -> None:
    dtx: float = dt * (N - 2)
    dty: float = dt * (N - 2)

    for j in range(1, N - 1):
        for i in range(1, N - 1):
            tmp1 = dtx * velocX[IX(i, j)]
            tmp2 = dty * velocY[IX(i, j)]
            x = i - tmp1
            y = j - tmp2

            if x < 0.5:
                x = 0.5
            if x > N + 0.5:
                x = N + 0.5
            i0 = math.floor(x)
            i1 = i0 + 1.0

            if y < 0.5:
                y = 0.5
            if y > N + 0.5:
                y = N + 0.5
            j0 = math.floor(y)
            j1 = j0 + 1.0

            s1 = x - i0
            s0 = 1.0 - s1
            t1 = y - j0
            t0 = 1.0 - t1

            i0i = int(i0)
            i1i = int(i1)
            j0i = int(j0)
            j1i = int(j1)

            d[IX(i, j)] = s0 * (t0 * d0[IX(i0i, j0i)] + t1 * d0[IX(i0i, j1i)]) + s1 * (
                t0 * d0[IX(i1i, j0i)] + t1 * d0[IX(i1i, j1i)]
            )

    set_bnd(b, d, N)


def set_bnd(b: int, x: list[float], N: int = Fluid.N) -> None:
    """a way to keep fluid from leaking out of the box"""

    for i in range(1, N - 1):
        x[IX(i, 0)] = -x[IX(i, 1)] if b == 2 else x[IX(i, 1)]
        x[IX(i, N - 1)] = -x[IX(i, N - 2)] if b == 2 else x[IX(i, N - 2)]
        x[IX(0, i)] = -x[IX(1, i)] if b == 1 else x[IX(1, i)]
        x[IX(N - 1, i)] = -x[IX(N - 2, i)] if b == 1 else x[IX(N - 2, i)]

    x[IX(0, 0)] = 0.33 * (x[IX(1, 0)] + x[IX(0, 1)] + x[IX(0, 0)])
    x[IX(0, N - 1)] = 0.33 * (x[IX(1, N - 1)] + x[IX(0, N - 2)] + x[IX(0, N - 1)])
    x[IX(N - 1, 0)] = 0.33 * (x[IX(N - 2, 0)] + x[IX(N - 1, 1)] + x[IX(N - 1, 0)])
    x[IX(N - 1, N - 1)] = 0.33 * (
        x[IX(N - 2, N - 1)] + x[IX(N - 1, N - 2)] + x[IX(N - 1, N - 1)]
    )

File: fluid_simulation/go_impl/fluid.go
package main

import (
	"math"
)

const (
	N    int = 124
	iter int = 15
)

type Fluid struct {
	size int64
	dt   float64
	diff float64
	visc float64

	s       []float64
	density []float64

	Vx []float64
	Vy []float64
	Vz []float64

	Vx0 []float64
	Vy0 []float64
	Vz0 []float64
}

func step(f Fluid) {
	diffuse(1, f.Vx0, f.Vx, f.visc, f.dt)
	diffuse(2, f.Vy0, f.Vy, f.visc, f.dt)

	project(f.Vx0, f.Vy0, f.Vx, f.Vy)

	advect(1, f.Vx, f.Vx0, f.Vx0, f.Vy0, f.dt)
	advect(2, f.Vy, f.Vy0, f.Vx0, f.Vy0, f.dt)

	project(f.Vx, f.Vy, f.Vx0, f.Vy0)

	diffuse(0, f.s, f.density, f.diff, f.dt)
	advect(0, f.density, f.s, f.Vx, f.Vy, f.dt)

}

// Fluid methods
func addDensity(f Fluid, x int, y int, amount float64) {
	index := IX(x, y)
	f.density[index] += amount
}

func addVelocity(f Fluid, x int, y int, amountX float64, amountY float64) {
	index := IX(x, y)
	f.Vx[index] += amountX
	f.Vy[index] += amountY
}

// Fluid functions
func setBnd(b int, x []float64) {
	for i := 1; i < N-1; i++ {
		if b == 2 {
			x[IX(i, 0)] = -x[IX(i, 1)]
			x[IX(i, N-1)] = -x[IX(i, N-2)]
		} else {
			x[IX(i, 0)] = x[IX(i, 1)]
			x[IX(i, N-1)] = x[IX(i, N-2)]
		}
	}
	for j := 1; j < N-1; j++ {
		if b == 1 {
			x[IX(0, j)] = -x[IX(1, j)]
			x[IX(N-1, j)] = -x[IX(N-2, j)]
		} else {
			x[IX(0, j)] = x[IX(1, j)]
			x[IX(N-1, j)] = x[IX(N-2, j)]
		}
	}

	x[IX(0, 0)] = 0.5 * (x[IX(1, 0)] + x[IX(0, 1)])
	x[IX(0, N-1)] = 0.5 * (x[IX(1, N-1)] + x[IX(0, N-2)])
	x[IX(N-1, 0)] = 0.5 * (x[IX(N-2, 0)] + x[IX(N-1, 1)])
	x[IX(N-1, N-1)] = 0.5 * (x[IX(N-2, N-1)] + x[IX(N-1, N-2)])
}

func advect(b int, d []float64, d0 []float64, velocX []float64, velocY []float64, dt float64) {
	var i0, i1, j0, j1 float64

	var dtx float64 = dt * (float64(N) - 2)
	var dty float64 = dt * (float64(N) - 2)

	var s0, s1, t0, t1 float64
	var tmp1, tmp2, x, y float64

	for j := 1; j < N-1; j++ {
		for i := 1; i < N-1; i++ {

			tmp1 = dtx * velocX[IX(i, j)]
			tmp2 = dty * velocY[IX(i, j)]
			x = float64(i) - tmp1
			y = float64(j) - tmp2

			if x < 0.5 {
				x = 0.5
			}
			if x > float64(N)+0.5 {
				x = float64(N) + 0.5
			}
			i0 = math.Floor(x)
			i1 = i0 + 1.0

			if y < 0.5 {
				y = 0.5
			}
			if y > float64(N)+0.5 {
				y = float64(N) + 0.5
			}
			j0 = math.Floor(y)
			j1 = j0 + 1.0

			s1 = x - i0
			s0 = 1.0 - s1
			t1 = y - j0
			t0 = 1.0 - t1

			i0i := int(i0)
			i1i := int(i1)
			j0i := int(j0)
			j1i := int(j1)

			d[IX(i, j)] = s0*(t0*d0[IX(i0i, j0i)]) + (t1 * d0[IX(i0i, j1i)]) + s1*(t0*d0[IX(i1i, j0i)]) + (t1 * d0[IX(i1i, j1i)])

		}
	}
	setBnd(b, d)
}

func project(velocX []float64, velocY []float64, p []float64, div []float64) {
	for i := 1; i < N-1; i++ {
		for j := 1; j < N-1; j++ {
			div[IX(i, j)] = -0.5 * (velocX[IX(i+1, j)] - velocX[IX(i-1, j)] + velocY[IX(i, j+1)] - velocY[IX(i, j-1)]) / float64(N)
			p[IX(i, j)] = 0
		}
	}

	setBnd(0, div)
	setBnd(0, p)
	linSolve(0, p, div, 1, 6)
	for i := 1; i < N-1; i++ {
		for j := 1; j < N-1; j++ {
			velocX[IX(i, j)] -= 0.5 * (p[IX(i+1, j)] - p[IX(i-1, j)]) * float64(N)
			velocY[IX(i, j)] -= 0.5 * (p[IX(i, j+1)] - p[IX(i, j-1)]) * float64(N)
		}
	}
	setBnd(1, velocX)
	setBnd(2, velocY)

}

func linSolve(b int, x []float64, x0 []float64, a float64, c float64) {
	var cRecip float64 = 1.0 / c
	for k := 0; k < iter; k++ {
		for j := 1; j < N-1; j++ {
			for i := 1; i < N-1; i++ {
				x[IX(i, j)] = x0[IX(i, j)] + a*(x[IX(i+1, j)]+x[IX(i-1, j)]+x[IX(i, j+1)]+x[IX(i, j-1)])*cRecip
			}
		}
		setBnd(b, x)
	}

}

func diffuse(b int, x []float64, x0 []float64, diff float64, dt float64) {
	var a float64 = dt * diff * (float64(N) - 2) * (float64(N) - 2)
	linSolve(b, x, x0, a, 1+6*a)
}

// Helper functions
func IX(x int, y int) int {
	x = constrain(x, 0, N-1)
	y = constrain(x, 0, N-1)
	return x + y*N
}

func constrain[V int | float64](val V, min V, max V) V {
	if val > max {
		return max
	} else if val < min {
		return min
	} else {
		return val
	}
}

// Create an instance of fluid
func newFluid(size int64, dt float64, diff float64, visc float64) Fluid {
	fluid := Fluid{
		size: size,
		dt:   dt,
		diff: diff,
		visc: visc,

		s:       make([]float64, N*N),
		density: make([]float64, N*N),

		Vx: make([]float64, N*N),
		Vy: make([]float64, N*N),
		Vz: make([]float64, N*N),

		Vx0: make([]float64, N*N),
		Vy0: make([]float64, N*N),
		Vz0: make([]float64, N*N),
	}

	return fluid
}
